runtime/src/bank.rs

```diff
@@ -1959,6 +1959,7 @@ impl Bank {
     /// calculation and could shield other real accounts.
     pub fn verify_snapshot_bank(&self) -> bool {
         self.clean_accounts();
+        self.shrink_all_stale_slots();
```

Comment:
What is the performance of this?

Reply:
To be honest, I haven't measured it still, although I know this shrinking operation isn't trivial operation. I prioritized to land this for the poor devnet bullied by the break...

My general analysis (heh, not validated by measurements) are like this:

In both following cases, I think these performance cost is justifiable to avoid the DOS this PR is trying to address.

**the snapshot codepath (as you commented here)**: Given innocent snapshots, the shrinking process is basically no-op because almost all slots should be shrunken by the snapshot-producing validator already. The no-op means exactly that we're bailing out [rather early](https://github.com/solana-labs/solana/pull/9219/files#diff-2099c5256db4eb5975c8834af38f6456R914), after a new full-scan of all `AppendVec`s collectively. There should be no lock contention pressure because we're only user of AccountsDB at the time of snapshot load (=verification).

Given malicious snapshots, this adds linear time increase to reject them, ultimately bounded by `MAX_SNAPSHOT_ARCHIVE_UNPACKED_SIZE`. (By the way, I noticed `generate_index` is quadratic in that regard while working on this. but this is a different issue...)

**the non-snapshot codepath (as an additional remark)**: As validator continuously do this at the rate of 1 slot/100ms cycling all alive slots, it should be rather rare to do actual shrinking. So this is like the snapshot code path for indivisual slots. Even if it did, that slot is rooted (no other writer) and the shrunken slot is old by definition there should be no lock contention.

*Ultimately, this PR introduces 1 read of account data, amortized no read lock by batching under normal/healthy operation.* 

As a knob for performance tuning, we have [this magic number](https://github.com/solana-labs/solana/pull/9219/files#diff-2099c5256db4eb5975c8834af38f6456R908) to reduce the count of written accounts.

Also, we could make the before-bail-out cost even cheaper by maintaining the all account count in append vec.

Anyway, I'm landing this after my last-minute local testing as this is prioritized to land as i said and I think [you're aware](https://github.com/solana-labs/solana/pull/9219#pullrequestreview-387384059). :)

Reply:
@sakridge ^^ Enjoy you reading :) (oops, I've just forgot to mention.)

Reply:
Thanks